README
ScottKit -- a Scott Adams adventure toolkit
Copyright (C) Mike Taylor <mike@miketaylor.org.uk>
Distributed under the GNU GPL, version 2 -- see file GPL-2

This is a Ruby program to compile, decompile and run adventure games
in Scott Adams format, written as an exercise in Ruby rather than with
any great expectation that it will be useful.


RUNNING SCOTTKIT

The three modes of running are:
$ scottkit -c game.sck > game.sao  # compile
$ scottkit game.sao  # play the game; or use scottfree game.sao
$ scottkit -d game.sao  # decompile -- useful for cheating

Other command-line arguments enable wizard commands, load saved games,
set random seeds and enable various kinds of debugging output: run
scottkit -h for details.


TESTING

Unit testing can be done as follows:

	$ ruby1.9 -I lib -rrake/rake_test_loader test/test_*
	Loaded suite test/test_canonicalise
	Started
	...........
	Finished in 0.769568 seconds.

	11 tests, 37 assertions, 0 failures, 0 errors, 0 skips

I have verified that ScottKit can be used to play and win Scott
Adams's games #1, 2, and 4: Adventureland, Pirate Adventure and Voodoo
Castle.  I welcome reports of its being used to play and win other
games, or failing that, reports of how it fails.


GAME FORMAT

The file notes/Definition.scottfree-1.14 is taken from the ScottFree
package, which contains another interpreter for Scott Adams games.  I
got this file from release 1.14-9 of ScottFree.  notes/Definition is my
modified version of this file.  notes/Definition.saved-game is my own
analysis of the format of saved games from ScottFree.


GAMES

The directory "data" contains game files:
  adams -- Scott Adams's classic games (see Makefile for details)
  howarth -- Brian Howarth's  games (see Makefile for details)
  test -- games files used by unit-test suite
  tutorial -- tiny game used in ScottKit tutorial
  crystal -- Crystal of Chaos, a game written to exercise ScottKit
  dan-and-matt.sck -- game written by my two eldest sons


BUG TOLERANCE

-b (--bug-tolerant) option is needed for Adventure 14b (Buckaroo
Banzai), which places one item (the jug of jet fuel) in room 50 of 36.
This is mapped to 0 (not in play), which is presumably what was
intended.  None of Scott's own games is so sloppy.




Changes
Revision history for Ruby gem ScottKit.

0.4.0  Mon Mar  1 22:20:09 GMT 2010
* Move towards using YARD for documentation instead of Rdoc.  This
  includes changes to things like the format of the Changes file.  Far
  from perfect so far: for example, plain-text documentation like
  doc/Definition gets mangled.

0.3.0  February 28, 2010
* Move ScottKit out of its sub-siubdirectory in the 9Gb "mike"
  module into its own git module.
* Move repository hosting to github.
* Arrange for formatted documentation to appear at
  http://rdoc.info/projects/MikeTaylor/scottkit

0.2.0  February 28, 2010
* Add -z option to sleep after finishing -- useful when running in a
  DOS Shell under Windows, so that error messages can be seen.
* Add facility to load a game while playing using "#load"

0.1.0  February 28, 2010
* Add -p option to compile and play a game from source.

0.0.0  February 28, 2010
* Initial release.

STILL TO DO
* Tweak compiler to break actions when they need too many arguments as
  well as when they need too many instructions.
* Write tutorial scaffolding.
* Write other documentation, e.g. reference guide.
* Ensure that the right files are built by rdoc/YARD
* Write blog entry about ScottKit
* Maybe modify to run against GLK using the Glkx driver



test/test_play.rb
require 'test/unit'
require 'scottkit/game'
require 'stringio'
require 'scottkit/withio'

class TestPlay < Test::Unit::TestCase #:nodoc:
  def test_play_tutorial7; play("t7"); end
  def test_play_crystal; play("crystal"); end

  def play(name)
    game = ScottKit::Game.new({ :read_file =>
                                "data/test/#{name}.solution",
                                :random_seed => 12368, :no_wait => true })
    game.load(IO.read("data/test/#{name}.sao"))
    f = StringIO.new
    withIO(nil, f) { game.play }
    expected = File.read("data/test/#{name}.transcript")
    assert_equal(expected, f.string)
  end
end


test/test_save.rb
require 'test/unit'
require 'scottkit/game'
require 'stringio'
require 'scottkit/withio'

class TestSave < Test::Unit::TestCase #:nodoc:
  # Can't use a setup() method here as the two test-cases need the
  # games to be initialised with different options.

  def test_save_crystal
    game = ScottKit::Game.new({ :random_seed => 12368, :echo_input => true })
    game.load(IO.read("data/test/crystal.sao"))
    withIO(File.new("data/test/crystal.save-script"), 
           File.new("/dev/null", "w")) do
      game.play()
    end
    assert_equal(File.read("TMP"), File.read("data/test/crystal.save-file"))
    File.unlink "TMP"
  end

  def test_resave_crystal
    game = ScottKit::Game.new({ :random_seed => 12368, :echo_input => true,
        :restore_file => "data/test/crystal.save-file" })
    game.load(IO.read("data/test/crystal.sao"))
    withIO(StringIO.new("save game\nTMP"), File.new("/dev/null", "w")) do
      game.play()
    end
    assert_equal(File.read("TMP"), File.read("data/test/crystal.save-file"))
    File.unlink "TMP"
  end
end


test/withio_test.rb
#!/usr/local/bin/ruby1.9 -w

# Run as: echo Bart Lisa | ./withio_test.rb && echo == && cat /tmp/xcv
# Expecting the output:
#	first line is 'Bart Lisa'
#	saved line 'Eric'
#	died in withIO: divided by 0
#	saved line was 'Eric'
#	==
#	first line was 'Bart Lisa'
#	oldout is of class IO

require 'stringio'
require 'scottkit/withio'

saved = nil
begin
  line = gets
  puts "first line is '#{line.chomp}'"
  withIO(StringIO.new("Eric\nthe"), File.new("/tmp/xcv", "w")) do
    |oldin, oldout|
    puts "first line was '#{line.chomp}'"
    puts "oldout is of class #{oldout.class}"
    saved = gets
    oldout.puts "saved line '#{saved.chomp}'"
    puts 1/0
  end
rescue Exception => e
  puts "died in withIO: #{e}"
end
puts "saved line was '#{saved.chomp}'"


lib/scottkit/decompile.rb
module ScottKit
  class Game
    private

    def quote(token) #:nodoc:
      if ((token !~ /^([!a-z_0-9-]+)$/i || token =~ /\n/) ||
          Compiler::Lexer::TOKENMAP[token])
        "\"#{token.gsub(/[""]/, '\'')}\""
      else
        token
      end
    end

    def decompile(f)
      f << "# #{@rooms.size} rooms, "
      f << "#{@items.size} items, "
      f << "#{@actions.size} actions\n"
      f << "# #{@messages.size} messages, "
      f << "#{defined?(@ntreasures) ? @ntreasures : "UNDEFINED"} treasures, "
      f << "#{@verbs.size} verbs/nouns\n"
      f.puts "ident #{@id}" if defined? @id
      f.puts "version #{@version}" if defined? @version
      f.puts "wordlen #{@wordlen}" if defined? @wordlen
      f.puts "maxload #{@maxload}" if defined? @maxload
      f.puts "lighttime #{@lamptime}" if defined? @lamptime
      f.puts "unknown1 #{@unknown1}" if defined? @unknown1
      f.puts "unknown2 #{@unknown2}" if defined? @unknown2
      f.puts "start #{quote roomname @startloc}" if defined? @startloc
      ### Do NOT change the nested if's to a single &&ed one: for
        # reasons that I do not at all understand, doing so results in
        # the protected statement being executed when @treasury is 0
      if defined? @treasury
        if @treasury != 0
          f.puts "treasury #{quote roomname @treasury}" 
        end
      end
      f.puts
      decompile_wordgroup(f, @verbs, "verb")
      decompile_wordgroup(f, @nouns, "noun")

      @rooms.each.with_index do |room, i|
        next if i == 0
        f.puts "room " << quote(roomname(i)) << " \"#{room.desc}\""
        room.exits.each.with_index do |exit, j|
          if exit != 0
            f.puts "\texit #{dirname(j)} #{quote roomname(exit)}"
          end
        end
        f.puts
      end

      @items.each.with_index do |item, i|
        f.puts "item #{quote itemname(i)} \"#{item.desc}\""
        f.puts "\tcalled #{quote item.name}" if item.name
        f.puts case item.startloc
               when ROOM_CARRIED then "\tcarried"
               when ROOM_NOWHERE then "\tnowhere"
               else "\tat #{quote roomname(item.startloc)}"
               end
        f.puts
      end

      @actions.each { |action| action.decompile(f) }
    end

    def decompile_wordgroup(f, list, label)
      canonical = nil
      synonyms = []
      printed = false

      list.each.with_index do |word, i|
        if (word =~ /^\*/)
          synonyms << word.sub(/^\*/, "")
        end
        if (word !~ /^\*/ || i == list.size-1)
          if synonyms.size > 0
            f.print "#{label}group #{quote canonical} "
            f.puts synonyms.map { |token| quote token }.join(" ")
            printed = true
          end
          canonical = word
          synonyms = []
        end
      end

      f.puts if printed
    end

    public :decompile # Must be visible to driver program
    public :quote # Needed for contained classes' decompile()/render() methods


    class Action
      def quote(*args); @game.quote(*args); end

      def decompile(f)
        emitted_noun_or_condition = false
        if self.verb == 0 then
          f << "occur"
          f << " " << self.noun << "%" if self.noun != 100
        else
          f << "action #{quote @game.verbs[self.verb]}"
          if self.noun != 0
            f << " #{quote @game.nouns[self.noun]}"
            emitted_noun_or_condition = true
          end
        end
        self.conds.each.with_index do |cond, i|
          f << (i == 0 ? " when " : " and ") << cond.render
          emitted_noun_or_condition = true
        end
        f << ":" if self.verb != 0 && !emitted_noun_or_condition
        f.puts
        args = @args.clone
        self.instructions.each do |instruction|
          f.puts "\t" + instruction.render(args)
        end
        if (self.comment != "")
          f.puts "\tcomment \"#{self.comment}\""
        end
        f.puts
      end
    end

    class Condition
      def quote(*args); @game.quote(*args); end

      def render
        type = OPS[@cond][1]
        res = quote(OPS[@cond][0])
        res += " " +
          quote(type == :room ? @game.roomname(@value) :
                type == :item ? @game.itemname(@value) :
                type == :number ? String(@value) : "ERROR") if
          type != :NONE
        res
      end
    end

    class Instruction
      def quote(*args); @game.quote(*args); end

      def render(args)
        if (@op == 0)
          return "NOP" # shouldn't happen
        elsif (@op <= 51)
          return "print #{quote @game.messages[@op]}"
        elsif (@op >= 102)
          return "print #{quote @game.messages[@op-50]}"
        end

        op = OPS[@op-52]
        return "UNKNOWN_OP" if !op
        op[0] + case op[1]
        when :item
          " #{quote @game.itemname(args.shift)}"
        when :room
          " #{quote @game.roomname(args.shift)}"
        when :number
          " #{@game.quote String(args.shift)}"
        when :item_item
          " #{quote @game.itemname(args.shift)}" +
          " #{quote @game.itemname(args.shift)}"
        when :item_room
          " #{quote @game.itemname(args.shift)}" +
          " #{quote @game.roomname(args.shift)}"
        when :NONE
          "" # Nothing to add
        else
          " UNKNOWN_PARAM"
        end
      end
    end
  end
end


lib/scottkit/compile.rb
# -*- coding: utf-8 -*-
# Test with: cd /usr/local/src/mike/scott/scottkit && ruby1.9 -I lib bin/scottkit -c data/tutorial/t7.sck

require 'pp'

module ScottKit
  class Game
    class Compiler
      private

      # Creates a new compiler for the specified game, set up ready to
      # compile the game in the specified file.
      #
      # The input file may be specified either as a filename or a
      # filehandle, or both.  If both are given, then the filename is
      # used only in reporting to help locate errors.  _Some_ value
      # must be given for the filename: an empty string is OK.
      #
      # (In case you're wondering, the main reason this has to be
      # passed a Game object is that the behaviour of compile is
      # influenced by the game's options.)
      #
      def initialize(game, filename, fh = nil)
        @game = game
        @lexer = Lexer.new(game, filename, fh)
      end

      # Compiles the specified game-source file, writing the resulting
      # object file to stdout, whence it should be redirected into a
      # file so that it can be played.  Yes, this API is sucky: it
      # would be better if we had a simple compile method that builds
      # the game in memory in a form that can by played, and which can
      # then also be saved as an object file by some other method --
      # but that would have been more work for little gain.
      #
      def compile_to_stdout
        begin
          tree = parse
          generate_code(tree)
          true
        rescue
          return false if String($!) == "syntax error"
          raise
        end
      end

      def parse
        ident, version, unknown1, unknown2, start, treasury, maxload,
        wordlen, lighttime, lightsource =
          nil, nil, nil, nil, nil, nil, nil, nil, nil, nil
        rooms = [ CRoom.new(nil, nil, {}) ]
        items, actions, verbgroups, noungroups = [], [], [], []

        while peek != :eof
          case peek
          when :ident then skip; ident = match :symbol
          when :version then skip; version = match :symbol
          when :unknown1 then skip; unknown1 = match :symbol
          when :unknown2 then skip; unknown2 = match :symbol
          when :start then skip; start = match :symbol
          when :treasury then skip; treasury = match :symbol
          when :maxload then skip; maxload = match :symbol
          when :wordlen then skip; wordlen = match :symbol
          when :lighttime then skip; lighttime = match :symbol
          when :lightsource then skip; lightsource = match :symbol
          when :room then rooms << parse_room
          when :item then items << parse_item(rooms.size-1)
          when :action then actions << parse_action
          when :occur then actions << parse_occur
          when :verbgroup then verbgroups << parse_wordgroup
          when :noungroup then noungroups << parse_wordgroup
          else match nil, "new directive"
          end
        end

        if peek != :eof
          error "additional text remains after parsing"
        end

        CGame.new(ident, version, unknown1, unknown2, start, treasury,
                  maxload, wordlen, lighttime, lightsource, rooms,
                  items, actions, verbgroups, noungroups)
      end

      def parse_room
        match :room
        name = match :symbol
        desc = match :symbol
        exits = {}
        while peek == :exit
          match :exit
          direction = match :direction
          dest = match :symbol
          exits[direction] = dest
        end
        CRoom.new(name, desc, exits)
      end

      def parse_item(last_room)
        match :item
        name = match :symbol
        desc = match :symbol
        called, where = nil, nil
        while true
          case peek
          when :called then match :called; called = match :symbol
          when :at then match :at; where = match :symbol
          when :nowhere then match :nowhere; where = ROOM_NOWHERE
          when :carried then match :carried; where = ROOM_CARRIED
          else break
          end
        end
        CItem.new(name, desc, called, where ? where : last_room)
      end

      def parse_action
        match :action
        verb = match :symbol
        if peek == :when
          noun = nil
        elsif peek == ":" # to terminate single-word actions if no conditions
          skip
          noun = nil
        else
          noun = match :symbol
          skip if peek == ":" # optional
        end

        conds, instructions, comment = parse_actionbody
        CAction.new(verb, noun, conds, instructions, comment)
      end

      def parse_occur
        match :occur
        if peek == :percent
          chance = match :percent
          skip if peek == ":" # optional
        else
          chance = nil
        end

        conds, instructions, comment = parse_actionbody
        CAction.new(nil, chance, conds, instructions, comment)        
      end

      def parse_actionbody
        conds = []
        while peek == :when || peek == :and
          skip
          op = match :symbol
          type = Condition::OPStotype[op] or
            error "unknown condition op '#{op}'"
          case type
          when :NONE then val = 0 # Any numeric value is fine here
          when :number then val = match :symbol
          when :room then val = match :symbol
          when :item then val = match :symbol
          else error "condition op '#{op}' has unknown type '#{type}'"
          end
          conds << [ op, val ]
        end

        instructions = []
        while peek == :symbol
          op = match :symbol
          if op == "print"
            val = [ match(:symbol) ]
          else
            type = Instruction::OPStotype[op] or
              error "unknown instruction op '#{op}'"
            case type
            when :NONE then val = []
            when :number then val = [ match(:symbol) ]
            when :room then val = [ match(:symbol) ]
            when :item then val = [ match(:symbol) ]
            when :item_item then val = [ match(:symbol), match(:symbol) ]
            when :item_room then val = [ match(:symbol), match(:symbol) ]
            else error "instruction op '#{op}' has unknown type '#{type}'"
            end
          end
          instructions << [ op, val ]
        end

        comment = nil
        if peek == :comment
          skip
          comment = match :symbol
        end

        [ conds, instructions, comment ]
      end

      def parse_wordgroup
        skip
        words = []
        while peek == :symbol
          words << match(:symbol)
        end
        words
      end

      # Skip over the current token
      def skip; match peek; end

      # Delegators through to the lexer class: just to keep parser source terse
      def peek; @lexer.peek; end
      def match(token, estr = nil); @lexer.match token, estr; end
      def error(str); @lexer.error str; end


      def generate_code(tree)
        @had_errors = false
        rooms = tree.rooms
        items = tree.items

        if tree.lightsource then
          # The light-source is always item #9, so swap as necessary
          lindex = items.index { |x| x.name == tree.lightsource } or
            gerror "lightsource '#{tree.lightsource}' does not exist"
          if (lindex != ITEM_LAMP)
            items << CItem.new(nil, "", nil, 0) while items.size <= ITEM_LAMP
            items[lindex], items[ITEM_LAMP] = items[ITEM_LAMP], items[lindex]
          end
        end

        # Make name->index maps for rooms and items
        roommap = { "_ROOM0" => 0 }
        itemmap = {}
        rooms.each.with_index { |room, index| roommap[room.name] = index }
        items.each.with_index { |item, index| itemmap[item.name] = index }

        startindex, treasuryindex = [ [ tree.start, "start" ],
                                      [ tree.treasury, "treasury" ]
                                      ].map {
          |ref| loc, caption = *ref
          !loc ? 1 : roommap[loc] or
            gerror "#{caption} room '#{loc}' does not exist"
        }

        # Resolve room names in exits
        rooms.each do |room|
          room.exits.each do |dir, dest|
            room.exits[dir] = roommap[dest] or
              gerror "'#{dest}' (#{dir} from #{room.name}) does not exist"
          end
        end

        # Resolve room names in item locations
        items.each do |item|
          if item.where.class == String
            num = room_by_name(item.where, roommap) or
              gerror "location '#{item.where}' for #{item.desc}) does not exist"
            item.where = num
          end
        end

        # Map each verb and noun to group of all its synonyms
        @wordlen = Integer(tree.wordlen ||= 3)
        verbtogroup, nountogroup = [ tree.verbgroups,
                                     tree.noungroups ].map { |groups|
          groups = groups.map { |list| list.map { |word| word.upcase[0, @wordlen] } }
          res = {}
          groups.each do |list|
            list.each { |word| res[word] = list }
          end
          res
        }

        # Compile vocabulary, including synonyms.
        verbs = [ "AUT" ]
        nouns = [ "ANY" ]
        verbmap, nounmap = {}, {}

        # Verb 1 is GO, verb 10 is GET, verb 18 is DROP (always).
        [ ["go", 1], ["get", 10], ["drop", 18] ]. each do |pair|
          insert_word(verbtogroup, verbs, verbmap, *pair)
        end
        # Nouns 1-6 are directions: no synonyms possible
        0.upto(5).each do |i|
          insert_word(nountogroup, nouns, nounmap, @game.dirname(i), i+1)
        end


        # Messages from actions will be accumulated here
        messages = [ "" ]       # Maps message-number to message
        messagemap = {}         # Maps message to message-number

        # Instructions must not exceed four per batch
        actions = []
        tree.actions.each do |action|
          ins, acc = action.instructions, []
          while ins.size > 4
            acc.concat ins.shift(3)
            acc.push [ "continue" ]
          end
          acc.concat ins
          acc.push [0] while acc.size % 4 != 0

          # We now have batches of four instructions; each but the
          # first must be placed in a new action.
          action.instructions = acc.shift(4)
          actions << action
          actions << CAction.new(nil, 0, [], acc.shift(4), "cont", []) while
            acc.count != 0
        end

        # Resolve room and item names in actions and occurrences
        actions.each do |action|
          if action.verb
            # Actual actions
            action.verb = insert_word(verbtogroup, verbs, verbmap, action.verb)
            if action.noun
              action.noun = insert_word(nountogroup, nouns, nounmap,
                                        action.noun)
            else
              action.noun = 0
            end
          else
            # Occurrences
            action.verb = 0
            action.noun = Integer(action.noun || 100)
          end

          action.conds.each do |cond|
            op, arg = cond[0], cond[1]
            opcode = Condition::OPStoindex[op]
            type = Condition::OPStotype[op]
            raise "impossible unknown condition op '#{op}'" if
              !opcode || !type
            cond[0] = opcode
            case type
            when :NONE then # nothing to do
            when :number then cond[1] = Integer(cond[1])
            when :room then cond[1] = roommap[arg] or
                gerror "unknown room in condition '#{arg}'"
            when :item then cond[1] = itemmap[arg] or
                gerror "unknown item in condition '#{arg}'"
            else gerror "condition op '#{op}' has unknown type '#{type}'"
            end
          end

          gathered_args = []
          action.instructions.each do |ins|
            op, args = ins[0], ins[1]
            arg0, arg1 = *args
            if op == 0
              next
            elsif op == "print"
              if !(msgno = messagemap[arg0])
                messages << arg0
                msgno = messagemap[arg0] = messages.size-1
              end
              ins[0] = msgno <= 51 ? msgno : msgno+50
              next
            end
            opcode = Instruction::OPStoindex[op]
            type = Instruction::OPStotype[op]
            raise "impossible unknown instruction op '#{op}'" if
              !opcode || !type
            ins[0] = opcode
            case type
            when :NONE then # nothing to do
            when :number then gathered_args << Integer(arg0)
            when :room then gathered_args << (roommap[arg0] or
                gerror "unknown room in instruction '#{arg0}'")
            when :item then gathered_args << (itemmap[arg0] or
                gerror "unknown item in instruction '#{arg0}'")
            when :item_item then gathered_args << (itemmap[arg0] or
                gerror "unknown item in instruction '#{arg0}'")
              gathered_args << (itemmap[arg1] or
                gerror "unknown item in instruction '#{arg1}'")
            when :item_room then gathered_args << (itemmap[arg0] or
                gerror "unknown item in instruction '#{arg0}'")
              gathered_args << (roommap[arg1] or
                gerror "unknown room in instruction '#{arg1}'")
            else gerror "instruction op '#{op}' has unknown type '#{type}'"
            end
          end
          action.gathered_args = gathered_args
        end

        # Add auto-get names of items to vocabulary
        items.each do |item|
          insert_word(nountogroup, nouns, nounmap, item.called) if item.called
        end

        1.upto([ verbs.size-1, nouns.size-1 ].max) do |i|
          verbs[i] = "" if !verbs[i]
          nouns[i] = "" if !nouns[i]
        end

        return if @had_errors

        # Write header
        puts tree.unknown1 || 0
        puts items.size-1
        puts actions.size-1
        puts verbs.size-1
        puts rooms.size-1
        puts tree.maxload || -1
        puts startindex
        puts items.select { |x| x.desc[0] == "*" }.count
        puts tree.wordlen
        puts tree.lighttime || -1
        puts messages.size-1
        puts treasuryindex
        puts # Blank line

        # Actions
        actions.each do |action|
          print 150*action.verb + action.noun, " "

          print action.conds.map { |x| String(x[0] + 20 * x[1]) + " " }.join
          print action.gathered_args.map { |x| String(20*x) + " " }.join
          nconds = action.conds.size + action.gathered_args.size
          raise "condition has #{nconds} conditions" if nconds > 5
          (5-nconds).times { print "0 " }

          ins = action.instructions.map { |x| x[0] }
          (4-ins.count).times { ins << 0 }
          puts "#{150*ins[0] + ins[1]} #{150*ins[2] + ins[3]}\n"
        end
        puts # Blank line

        # Vocab
        verbs.each.with_index do |verb, i|
          puts "\"#{verb}\" \"#{nouns[i]}\""
        end
        puts # Blank line

        # Rooms
        rooms.each do |room|
          0.upto(5).each do |i|
            exit = room.exits[@game.dirname(i)]
            print(exit ? exit : 0, " ")
          end
          print "\"#{room.desc}\"\n"
        end
        puts # Blank line

        # Messages
        messages.each do |message|
          puts "\"#{message}\"\n"
        end
        puts # Blank line

        # Items
        items.each do |item|
          desc = item.desc
          desc += "/" + item.called.upcase[0, @wordlen] + "/" if item.called
          puts "\"#{desc}\" #{item.where}"
        end
        puts # Blank line

        # Action comments
        actions.each do |action|
          puts "\"#{action.comment || ""}\"\n"
        end
        puts # Blank line

        # Trailer
        puts tree.version || 0
        puts tree.ident || 0
        puts tree.unknown2 || 0
      end

      def room_by_name(loc, roommap)
        if @game.options[:bug_tolerant] && loc[0,5] == "_ROOM"
          Integer(loc[5,999])
        else
          roommap[loc]
        end
      end

      # Complex API here, sorry.  If word, or an equivalent word
      # according to synmap, is not already in list and map, inserts
      # it and its synonyms into both, with map hashing words to the
      # index they appear at.  Word is inserted at index if specified
      # and otherwise at the first free slot, or off the end if there
      # are no free slots.  Synonyms, if any, follow thereafter in a
      # block.  Returns the index of the word in list
      def insert_word(synmap, list, map, word, index = nil)
        word = word.upcase[0, @wordlen]
        syn = synmap[word] || [word]
        canonical = syn[0]
        return map[canonical] if map[canonical]
        if !index
          index = 1
          index += 1 while syn.each_index.any? { |i| list[index+i] != nil }
        end

        firstindex = index
        syn.each do |thisword|
          if list[index]
            return(gerror "can't insert word '#{thisword}' " +
                   "@at position #{index} -- got '#{list[index]}'")
          end
          list[index] = index == firstindex ? thisword : "*"+thisword
          map[thisword] = firstindex
          index += 1
        end
        firstindex
      end

      def gerror(str)
        $stderr.puts "error: #{str}"
        @had_errors = true
        0
      end


      public :compile_to_stdout # Must be visible to Game.compile()
      public :parse # Used by test_compile.rb


      CGame = Struct.new(:ident, :version, :unknown1, :unknown2,
                         :start, :treasury, :maxload, :wordlen,
                         :lighttime, :lightsource, :rooms, :items,
                         :actions, :verbgroups, :noungroups) #@private
      CRoom = Struct.new(:name, :desc, :exits) #@private
      #@private
      CItem = Struct.new(:name, :desc, :called, :where)
      # @private
      CAction = Struct.new(:verb, :noun, :conds, :instructions,
                           :comment, :gathered_args)


      class Lexer #:nodoc:
        attr_reader :lexeme

        TOKENMAP = {
          "start" => :start,
          "treasury" => :treasury,
          "ident" => :ident,
          "version" => :version,
          "unknown1" => :unknown1,
          "unknown2" => :unknown2,
          "maxload" => :maxload,
          "wordlen" => :wordlen,
          "lighttime" => :lighttime,
          "lightsource" => :lightsource,
          "room" => :room,
          "exit" => :exit,
          "north" => :direction, "south" => :direction, "east" => :direction,
          "west" => :direction, "up" => :direction, "down" => :direction,
          "item" => :item,
          "called" => :called,
          "at" => :at,
          "nowhere" => :nowhere,
          "carried" => :carried,
          "action" => :action,
          "occur" => :occur,
          "when" => :when,
          "and" => :and,
          "comment" => :comment,
          "verbgroup" => :verbgroup,
          "noungroup" => :noungroup,
        }

        def initialize(game, filename, fh = nil)
          if !fh
            fh = File.new(filename)
          end
          @game, @filename, @fh = game, filename, fh
          @linenumber = 0
          @buffer = ""
          @lookahead = nil
        end

        def error(str)
          filename = (defined? @filename) ? @filename : "<UNKNOWN>"
          $stderr.puts "#{filename}:#{@linenumber}:#{str}"
          raise "syntax error"
        end

        def lex
          token = _lex
          @game.dputs :show_tokens, "token: #{render(token)}"
          token
        end

        def _lex
          @buffer.lstrip!
          while @buffer == "" do
            if !(@buffer = @fh.gets)
              return :eof
            end
            @linenumber += 1
            @buffer.chomp!
            @buffer.rstrip!
            @buffer.lstrip!
          end

          if @buffer[0] == "#"
            # Comment runs to end of line
            @buffer = ""
            return _lex # Be honest, a GOTO would be better here
          elsif match = @buffer.match(/^"(.*?)"/)
            @lexeme, @buffer = match[1], match.post_match
            :symbol
          elsif match = @buffer.match(/^"(.*)/)
            # Multi-line string -- can include hashes and indents
            s = match[1]
            while @buffer = @fh.gets
              @linenumber += 1
              @buffer.chomp!
              if match = @buffer.match(/(.*?)"/)
                @lexeme = s + "\n" + match[1]
                @buffer = match.post_match
                break
              else
                s += "\n" + @buffer
              end
            end
            :symbol
          elsif match = @buffer.match(/^(\d+)%/)
            @lexeme, @buffer = match[1], match.post_match
            :percent
          elsif match = @buffer.match(/^([!a-z_0-9-]+)/i)
            @lexeme, @buffer = match[1], match.post_match
            TOKENMAP[@lexeme] || :symbol
          else
            # Must be a single character
            @lexeme = @buffer[0]
            @buffer[0] = ""
            @lexeme
          end
        end

        def peek
          @lookahead ||= lex
        end

        def match(expected, estr = nil)
          token = peek
          @lookahead = nil
          if token != expected
            error("expected #{estr || expected}, got #{render(token)}" +
                  " (before `#{@buffer.lstrip}')")
          end
          @lexeme
        end

        def render(token)
          extra = render_lexeme(token, @lexeme)
          extra ? "#{token} #{extra}" : token
        end

        def render_lexeme(token, lexeme)
          if token == :direction
            lexeme
          elsif token == :symbol
            "\"#{lexeme}\""
          elsif token == :percent
            "'#{lexeme}%'"
          else
            nil
          end
        end
      end
    end
  end
end
